README.md
# react-native-spinkit-button [![npm version](https://badge.fury.io/js/react-native-spinkit-button.svg)](https://badge.fury.io/js/react-native-spinkit-button)

`React-Native` `Button`, with collection of animated loading indicators `(react-native-spinkit)`.

# Preview

![Example of react-native-spinkit-button](https://github.com/taimoorsultani/react-native-spinkit-button/blob/main/example.gif)

## Getting Started

### Using npm

```bash
npm i react-native-spinkit-button
```

### Using yarn

```bash
yarn add react-native-spinkit-button
```

## Usage

### Text Only

```javascript
import React, {useState} from 'react';
import SpinkitButton from 'react-native-spinkit-button';

const App = () => {
  const [onlyTextLoading, setOnlyTextLoading] = useState(false);

  const onTextClick = () => {
    setOnlyTextLoading(true);
    setTimeout(() => {
      setOnlyTextLoading(false);
    }, 3000);
  };

  return (
    <SpinkitButton
      width={270}
      height={40}
      borderRadius={11}
      onPress={onTextClick}
      buttonStyle={styles.button}
      label={'ONLY TEXT'}
      labelStyle={styles.textButtonStyle}
      loading={onlyTextLoading}
      size={15}
      type={'CircleFlip'}
      color={'#FFFFFF'}
      animationDuration={300}
    />
  );
};
```

### Icon & Text

```javascript
import React, {useState} from 'react';
import SpinkitButton from 'react-native-spinkit-button';
import Icon from 'react-native-vector-icons/dist/FontAwesome';

const App = () => {
  const [iconAndTextLoading, setIconAndTextLoading] = useState(false);

  const iconWithTextClick = () => {
    setIconAndTextLoading(true);
    setTimeout(() => {
      setIconAndTextLoading(false);
    }, 3000);
  };

  return (
    <SpinkitButton
      width={270}
      height={40}
      borderRadius={11}
      onPress={iconWithTextClick}
      buttonStyle={styles.button}
      label={'TEXT with ICON'}
      labelStyle={styles.textButtonStyle}
      loading={iconAndTextLoading}
      labelAndTextContainer={styles.labelAndTextContainer}
      iconComponent={
        <Icon name="rocket" size={20} color="#FFFFFF" style={styles.icon} />
      }
      size={15}
      type={'Bounce'}
      color={'#FFFFFF'}
      animationDuration={300}
    />
  );
};
```

### Icon only

```javascript
import React, {useState} from 'react';
import SpinkitButton from 'react-native-spinkit-button';
import Icon from 'react-native-vector-icons/dist/FontAwesome';

const App = () => {
  const [onlyIconLoading, setOnlyIconLoading] = useState(false);

  const onIconClick = () => {
    setOnlyIconLoading(true);
    setTimeout(() => {
      setOnlyIconLoading(false);
    }, 3000);
  };

  return (
    <SpinkitButton
      width={270}
      height={40}
      borderRadius={11}
      onPress={onIconClick}
      buttonStyle={styles.button}
      labelStyle={styles.textButtonStyle}
      loading={onlyIconLoading}
      labelAndTextContainer={styles.labelAndTextContainer}
      iconComponent={
        <Icon name="rocket" size={20} color="#FFFFFF" style={styles.icon} />
      }
      size={15}
      type={'Wave'}
      color={'#FFFFFF'}
      animationDuration={300}
    />
  );
};
```

## Example

A full working example project is here [react-native-spinkit-button/example](https://github.com/taimoorsultani/react-native-spinkit-button/tree/main/example)

## Properties

| Prop                  |   Default    |     Type      | Description                                      |
| :-------------------- | :----------: | :-----------: | :----------------------------------------------- |
| width                 |    `250`     |   `number`    | width of button                                  |
| height                |     `40`     |   `number`    | height of button                                 |
| borderRadius          |     `10`     |   `number`    | border radius of button                          |
| buttonStyle           |     `{}`     |    `style`    | style of button                                  |
| onPress               |  `() => {}`  |  `function`   | the function to execute on tap                   |
| disabled              |   `false`    |    `bool`     | to disable the button                            |
| labelAndTextContainer |     `{}`     |    `style`    | label container styles                           |
| label                 |    `null`    |   `string`    | label of button                                  |
| labelStyle            |     `{}`     |    `style`    | style of label                                   |
| iconComponent         |    `null`    | `elementType` | `Icon` or `Image` component                      |
| loading               |   `false`    |    `bool`     | bool value to render spinkit indicator or button |
| size                  |     `10`     |   `number`    | size of spinkit indicator                        |
| type                  | `CircleFlip` |    `oneOf`    | type of spinkit indicator                        |
| color                 |  `#FFFFFF`   |   `string`    | color of spinkit indicator                       |
| animationDuration     |    `400`     |   `number`    | duration of animation                            |


### List of available spinkit indicator types

- CircleFlip
- Bounce
- Wave
- WanderingCubes
- Pulse
- ChasingDots
- ThreeBounce
- Circle
- 9CubeGrid
- WordPress (IOS only)
- FadingCircle
- FadingCircleAlt
- Arc (IOS only)
- ArcAlt (IOS only)

## Acknowledgment and Big Thanks to

[react-native-spinkit](https://github.com/maxs15/react-native-spinkit)

## License

This project is licensed under the MIT License -
(c) 2021 [Muhammad Taimoor Sultani](https://github.com/taimoorsultani), [MIT license](/LICENSE).


CHANGELOG.md
## 1.0.0 (Junine 1, 2021)

- Official Release!


src/index.js
import React, {useRef, useEffect} from 'react';
import PropTypes from 'prop-types';
import {TouchableOpacity, Animated, Text} from 'react-native';
import Spinner from 'react-native-spinkit';

import {buttonStyles} from './styles';

const SpinkitButton = props => {
  const {
    // Button Container
    width,
    height,
    borderRadius,
    buttonStyle,
    onPress,
    disabled,
    // Label and Text Container
    labelAndTextContainer,
    // Text
    label,
    labelStyle,
    // Icon
    iconComponent,
    // Spinner
    loading,
    size,
    type,
    color,
    animationDuration,
  } = props;

  const baseWidth = useRef(new Animated.Value(width)).current;

  const baseBorderRadius = useRef(new Animated.Value(borderRadius)).current;

  const baseTextOpacity = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    if (loading) {
      performAnimation(
        baseWidth,
        height,
        baseBorderRadius,
        height / 2,
        baseTextOpacity,
        0.1,
      );
    } else {
      performAnimation(
        baseWidth,
        width,
        baseBorderRadius,
        borderRadius,
        baseTextOpacity,
        1,
      );
    }
  }, [loading]);

  const performAnimation = (
    widthStart,
    widthEnd,
    borderRadiusStart,
    borderRadiusEnd,
    opacityStart,
    opacityEnd,
  ) => {
    Animated.timing(widthStart, {
      toValue: widthEnd,
      duration: animationDuration,
      useNativeDriver: false,
    }).start();
    Animated.timing(borderRadiusStart, {
      toValue: borderRadiusEnd,
      duration: animationDuration,
      useNativeDriver: false,
    }).start();
    Animated.timing(opacityStart, {
      toValue: opacityEnd,
      duration: animationDuration,
      useNativeDriver: false,
    }).start();
  };

  const renderTitle = () => {
    return (
      <Animated.View
        style={[
          {
            opacity: baseTextOpacity,
          },
          labelAndTextContainer && labelAndTextContainer,
        ]}>
        {iconComponent && iconComponent}
        {label && <Text style={labelStyle && labelStyle}>{label}</Text>}
      </Animated.View>
    );
  };

  const renderIndicator = () => {
    return <Spinner isVisible={true} color={color} size={size} type={type} />;
  };

  return (
    <TouchableOpacity
      disabled={disabled || loading}
      onPress={onPress}
      activeOpacity={1}>
      <Animated.View
        style={[
          buttonStyles.animatedContainer,
          buttonStyle && buttonStyle,
          {
            width: baseWidth,
            height: height,
            borderRadius: baseBorderRadius,
          },
        ]}>
        {loading ? renderIndicator() : renderTitle()}
      </Animated.View>
    </TouchableOpacity>
  );
};

SpinkitButton.prototype = {
  // Button Container
  width: PropTypes.number,
  height: PropTypes.number,
  borderRadius: PropTypes.number,
  buttonStyle: PropTypes.style,
  onPress: PropTypes.func,
  disabled: PropTypes.bool,
  // Label and Text Container
  labelAndTextContainer: PropTypes.style,
  // Text
  label: PropTypes.string,
  labelStyle: PropTypes.style,
  // Icon
  iconComponent: PropTypes.elementType,
  // Spinner
  loading: PropTypes.bool,
  size: PropTypes.number,
  type: PropTypes.oneOf([
    'CircleFlip',
    'Bounce',
    'Wave',
    'WanderingCubes',
    'Pulse',
    'ChasingDots',
    'ThreeBounce',
    'Circle',
    '9CubeGrid',
    'WordPress',
    'FadingCircle',
    'FadingCircleAlt',
    'Arc',
    'ArcAlt',
  ]),
  color: PropTypes.string,
  animationDuration: PropTypes.number,
};

SpinkitButton.defaultProps = {
  // Button Container
  width: 250,
  height: 40,
  borderRadius: 10,
  buttonStyle: {},
  onPress: () => {},
  disabled: false,
  // Label and Text Container
  labelAndTextContainer: {},
  // Text
  label: null,
  labelStyle: {},
  // Icon
  iconComponent: null,
  // Spinner
  loading: false,
  size: 10,
  type: 'CircleFlip',
  color: '#FFFFFF',
  animationDuration: 400,
};

export default SpinkitButton;


example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, {useState} from 'react';
import {ScrollView, StatusBar, StyleSheet, Text, View} from 'react-native';
import SpinkitButton from 'react-native-spinkit-button';
import Icon from 'react-native-vector-icons/dist/FontAwesome';

const App = () => {
  const [onlyTextLoading, setOnlyTextLoading] = useState(false);
  const [iconAndTextLoading, setIconAndTextLoading] = useState(false);
  const [onlyIconLoading, setOnlyIconLoading] = useState(false);

  const onlyTextClick = () => {
    setOnlyTextLoading(true);
    setTimeout(() => {
      setOnlyTextLoading(false);
    }, 3000);
  };

  const iconWithTextClick = () => {
    setIconAndTextLoading(true);
    setTimeout(() => {
      setIconAndTextLoading(false);
    }, 3000);
  };

  const onlyIconClick = () => {
    setOnlyIconLoading(true);
    setTimeout(() => {
      setOnlyIconLoading(false);
    }, 3000);
  };

  return (
    <View style={styles.main}>
      <StatusBar barStyle={'light-content'} />
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <View style={styles.innerContainer}>
          <Text style={styles.text}>React Native Spinkit Buttons</Text>
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={11}
            onPress={onlyTextClick}
            buttonStyle={[styles.base, styles.soild]}
            label={'ONLY TEXT'}
            labelStyle={styles.textButtonStyle}
            loading={onlyTextLoading}
            size={15}
            type={'CircleFlip'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={11}
            onPress={iconWithTextClick}
            buttonStyle={[styles.base]}
            label={'TEXT with ICON'}
            labelStyle={styles.textButtonStyle}
            loading={iconAndTextLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'Bounce'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={11}
            onPress={onlyIconClick}
            buttonStyle={[styles.base]}
            labelStyle={styles.textButtonStyle}
            loading={onlyIconLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'Wave'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={30}
            onPress={onlyTextClick}
            buttonStyle={[styles.base, styles.soild]}
            label={'ONLY TEXT'}
            labelStyle={styles.textButtonStyle}
            loading={onlyTextLoading}
            size={15}
            type={'WanderingCubes'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={30}
            onPress={iconWithTextClick}
            buttonStyle={[styles.base]}
            label={'TEXT with ICON'}
            labelStyle={styles.textButtonStyle}
            loading={iconAndTextLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'Pulse'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={30}
            onPress={onlyIconClick}
            buttonStyle={[styles.base]}
            labelStyle={styles.textButtonStyle}
            loading={onlyIconLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'ChasingDots'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={0}
            onPress={onlyTextClick}
            buttonStyle={[styles.base, styles.soild]}
            label={'ONLY TEXT'}
            labelStyle={styles.textButtonStyle}
            loading={onlyTextLoading}
            size={15}
            type={'ThreeBounce'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={0}
            onPress={iconWithTextClick}
            buttonStyle={[styles.base]}
            label={'TEXT with ICON'}
            labelStyle={styles.textButtonStyle}
            loading={iconAndTextLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'Circle'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
          <SpinkitButton
            width={270}
            height={40}
            borderRadius={0}
            onPress={onlyIconClick}
            buttonStyle={[styles.base]}
            labelStyle={styles.textButtonStyle}
            loading={onlyIconLoading}
            labelAndTextContainer={styles.labelAndTextContainer}
            iconComponent={
              <Icon
                name="rocket"
                size={20}
                color="#FFFFFF"
                style={styles.icon}
              />
            }
            size={15}
            type={'9CubeGrid'}
            color={'#FFFFFF'}
            animationDuration={300}
          />
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  main: {
    flex: 1,
    backgroundColor: '#303030',
    alignItems: 'center',
    paddingTop: 11,
    paddingBottom: 11,
  },
  innerContainer: {
    alignItems: 'center',
  },
  text: {
    color: '#FFFFFF',
    fontSize: 20,
    fontWeight: 'bold',
    textAlign: 'center',
    paddingBottom: 10,
  },
  base: {
    padding: 9,
    margin: 9,
    borderColor: 'rgba(208, 28, 98, 1)',
    borderWidth: 2,
    justifyContent: 'center',
    alignItems: 'center',
  },
  soild: {
    backgroundColor: '#D01C62',
    borderColor: 'rgba(208, 28, 98, 1)',
  },
  textButtonStyle: {
    color: '#FFFFFF',
  },
  labelAndTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    marginRight: 10,
    marginLeft: 10,
  },
});

export default App;
